api/rest/coordinator.go

```diff
@@ -0,0 +1,272 @@
+	statuses := make(chan bundleStatus)
+
+	for i := 0; i < numberOfWorkers; i++ {
+		go worker(ctx, jobs, statuses)
+	}
+
+	for _, n := range nodes {
+		logrus.WithField("IP", n.IP).Info("Sending creation request to node.")
+
+		// necessary to prevent the closure from giving the same node to all the calls
+		tmpNode := n
+		jobs <- func() bundleStatus {
+			select {
+			case <-ctx.Done():
+				return bundleStatus{
+					id:   id,
+					node: tmpNode,
+					err:  errors.New(contextDoneErrMsg),
+				}
+			default:
+			}
+
+			return c.createBundle(ctx, tmpNode, id, jobs)
+		}
+	}
+
+	return statuses
+}
+
+// CollectBundle waits until all the nodes' bundles have finished, downloads,
+// and merges them. The resulting bundle zip file path is returned.
+func (c parallelCoordinator) CollectBundle(ctx context.Context, bundleID string, numBundles int, statuses <-chan bundleStatus) (string, error) {
+
+	// holds the paths to the downloaded local bundles before merging
+	var bundlePaths []string
+
+	finishedBundles := 0
+	for finishedBundles < numBundles {
+		select {
+		case <-ctx.Done():
+			// TODO (https://jira.mesosphere.com/browse/DCOS_OSS-5303): this should be noted in the generated bundle and not just printed in the journal
+			logrus.WithField("ID", bundleID).Warn("context cancelled before all node bundles finished")
+			return mergeZips(bundleID, bundlePaths, c.workDir)
+		case s := <-statuses:
+			if !s.done {
+				logrus.WithError(s.err).WithField("IP", s.node.IP).WithField("ID", s.id).Info("Got status update. Bundle not ready.")
+				continue
+			}
+
+			// even if the bundle finished with an error, it's now finished so increment finishedBundles
+			finishedBundles++
+			if s.err != nil {
+				// TODO (https://jira.mesosphere.com/browse/DCOS_OSS-5303): this should be noted in the generated bundle and not just printed in the journal
+				logrus.WithError(s.err).WithField("IP", s.node.IP).WithField("ID", s.id).Warn("Bundle errored")
+				continue
+			}
+
+			bundlePath := filepath.Join(c.workDir, nodeBundleFilename(s.node))
+			err := c.client.GetFile(ctx, s.node.baseURL, s.id, bundlePath)
+			if err != nil {
+				// TODO (https://jira.mesosphere.com/browse/DCOS_OSS-5303): this should be noted in the generated bundle and not just printed in the journal
+				logrus.WithError(err).WithField("IP", s.node.IP).WithField("ID", s.id).Warn("Could not download file")
+				continue
+			}
+
+			bundlePaths = append(bundlePaths, bundlePath)
+		}
+	}
+
+	return mergeZips(bundleID, bundlePaths, c.workDir)
+}
+
+func mergeZips(bundleID string, bundlePaths []string, workDir string) (string, error) {
+
+	bundlePath := filepath.Join(workDir, fmt.Sprintf("bundle-%s.zip", bundleID))
+	mergedZip, err := os.Create(bundlePath)
+	if err != nil {
+		return "", err
+	}
+	defer mergedZip.Close()
+
+	zipWriter := zip.NewWriter(mergedZip)
+	defer zipWriter.Close()
+
+	for _, p := range bundlePaths {
+		err = appendToZip(zipWriter, p)
+		if err != nil {
+			return "", err
```

Comment:
Do we wan't to stop bundle creation if one bundle file is invalid? I think we should collect that error and add it to error summary

Reply:
You're probably right that we don't. I'll add a comment with that other ticket about the summary report.

Reply:
Add a TODO with ticket number